plugins/redfish/redfish-client.c

```diff
@@ -0,0 +1,598 @@
+		JsonObject *member = json_array_get_object_element (members, i);
+		if (!redfish_client_coldplug_member (self, member, error))
+			return FALSE;
+	}
+	return TRUE;
+}
+
+gboolean
+redfish_client_coldplug (RedfishClient *self, GError **error)
+{
+	JsonNode *node_root;
+	JsonObject *obj_root = NULL;
+	g_autoptr(GBytes) blob = NULL;
+	g_autoptr(JsonParser) parser = json_parser_new ();
+
+	/* nothing set */
+	if (self->update_uri_path == NULL) {
+		g_set_error_literal (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_INTERNAL,
+				     "no update_uri_path");
+		return FALSE;
+	}
+
+	/* try to connect */
+	blob = redfish_client_fetch_data (self, self->update_uri_path, error);
+	if (blob == NULL)
+		return FALSE;
+
+	/* get the update service */
+	if (!json_parser_load_from_data (parser,
+					 g_bytes_get_data (blob, NULL),
+					 (gssize) g_bytes_get_size (blob),
+					 error)) {
+		g_prefix_error (error, "failed to parse node: ");
+		return FALSE;
+	}
+	node_root = json_parser_get_root (parser);
+	if (node_root == NULL) {
+		g_set_error_literal (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_INVALID_FILE,
+				     "no root node");
+		return FALSE;
+	}
+	obj_root = json_node_get_object (node_root);
+	if (obj_root == NULL) {
+		g_set_error_literal (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_INVALID_FILE,
+				     "no root object");
+		return FALSE;
+	}
+	if (!json_object_get_boolean_member (obj_root, "ServiceEnabled")) {
+		g_set_error_literal (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_NOT_SUPPORTED,
+				     "service is not enabled");
+		return FALSE;
+	}
+	if (json_object_has_member (obj_root, "FirmwareInventory")) {
+		JsonObject *tmp = json_object_get_object_member (obj_root, "FirmwareInventory");
+		return redfish_client_coldplug_inventory (self, tmp, error);
+	}
+	if (json_object_has_member (obj_root, "SoftwareInventory")) {
+		JsonObject *tmp = json_object_get_object_member (obj_root, "SoftwareInventory");
+		return redfish_client_coldplug_inventory (self, tmp, error);
+	}
+	return TRUE;
+}
+
+static gboolean
+redfish_client_set_uefi_credentials (RedfishClient *self, GError **error)
+{
+	guint32 indications_le;
+	g_autofree gchar *userpass_safe = NULL;
+	g_auto(GStrv) split = NULL;
+	g_autoptr(GBytes) indications = NULL;
+	g_autoptr(GBytes) userpass = NULL;
+
+	/* get the uint32 specifying if there are EFI variables set */
+	indications = redfish_common_get_evivar_raw (REDFISH_EFI_INFORMATION_INDICATIONS,
+						     REDFISH_EFI_INFORMATION_GUID,
+						     error);
+	if (indications == NULL)
+		return FALSE;
+	if (g_bytes_get_size (indications) != 4) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_INVALID_FILE,
+			     "invalid value for %s, got %" G_GSIZE_FORMAT " bytes",
+			     REDFISH_EFI_INFORMATION_INDICATIONS,
+			     g_bytes_get_size (indications));
+		return FALSE;
+	}
+	memcpy (&indications_le, g_bytes_get_data (indications, NULL), 4);
```

Comment:
This approach might be problematic on different endian-ness.

Reply:
So, I wondered about this. I think UEFI is explicitly LE-only machines, and regardless I'm not actually sure if the value from the efivars sysfs dir is in host byte order or always just in LE...

Reply:
Yeah as far as I can tell s390x and ppc don't have UEFI implementations.

Reply:
@superm1 do you know how we'd do `-Denable-redfish=false` just for the Debian s390x case?

Reply:
@hughsie https://github.com/hughsie/fwupd/blob/master/contrib/ci/build_debian_s390x.sh is how the s390x build is kicked off.  Just add it to that list of options and document it being changed in https://github.com/hughsie/fwupd/tree/master/contrib/ci/README.md

Reply:
Works great, thanks.